core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java

```diff
@@ -1662,10 +1695,26 @@ private ServerConfig buildServerConfig(List<ServerPort> serverPorts) {
         final SslContext defaultSslContext = findDefaultSslContext(defaultVirtualHost, virtualHosts);
         final Collection<ServerPort> ports;
 
-        this.ports.forEach(
-                port -> checkState(port.protocols().stream().anyMatch(p -> p != PROXY),
-                                   "protocols: %s (expected: at least one %s or %s)",
-                                   port.protocols(), HTTP, HTTPS));
+        for (ServerPort port : this.ports) {
+            checkState(port.protocols().stream().anyMatch(p -> p != PROXY),
+                       "protocols: %s (expected: at least one %s or %s)",
+                       port.protocols(), HTTP, HTTPS);
+        }
+
+        // The port numbers of port-based virtual hosts must exist in 'ServerPort's.
+        final List<VirtualHost> portBasedVirtualHosts = virtualHosts.stream()
+                                                                    .filter(v -> v.port() > 0)
+                                                                    .collect(toImmutableList());
+        final List<Integer> portNumbers = this.ports.stream()
+                                                    .map(port -> port.localAddress().getPort())
+                                                    .filter(port -> port > 0)
+                                                    .collect(toImmutableList());
+        for (VirtualHost virtualHost : portBasedVirtualHosts) {
+            final int virtualHostPort = virtualHost.port();
+            final boolean portMatched = portNumbers.stream().anyMatch(port -> port == virtualHostPort);
```

Comment:
Can't we just change that the port is also set to the `ServerBuilder` when `ServerBuilder.virtualHost(port)` is called? so that users do not have to call `sb.http(x).virtualHost(x)`? 🤔

Reply:
I thought the port of a virtual host represents IP(L4) layer's port that does not belong to a specific protocol. 
When comes to TLS configuration, `ServerBuilder.virtualHost(tlsPort)` seems a bit vague.

Reply:
Yeah, that makes sense. Let's keep it as it is.